Refactor Notification::permission to follow the same code path as Permissions::query

RupinMittal opened the pull request:
#### 92a17b98d823519296e40f2440baa5af4e049c92
<pre>
Refactor Notification::permission to follow the same code path as Permissions::query
<a href="https://bugs.webkit.org/show_bug.cgi?id=243502">https://bugs.webkit.org/show_bug.cgi?id=243502</a>

Reviewed by NOBODY (OOPS!).

There are two functions for obtaining the permission state of the Notifications API.
We need to provide support for both to ensure compatibility with web standards. This
patch ensures that they follow the same code path since having two different paths is
unnecessary and hurts maintainability. This code path goes to the UIProcess and then
to the client to get the permission state.

This patch also removes the now uneeded cache in WebNotificationManager that
Notification::permission was relying on and enables the WebProcess cache in
WebPermissionController to be used.

* LayoutTests/http/tests/notifications/service-worker-notification-permission-expected.txt: Added.
* LayoutTests/http/tests/notifications/service-worker-notification-permission.html: Added.
* LayoutTests/http/tests/notifications/service-worker-notification-permission.js: Added.
(self.onmessage):
* LayoutTests/http/tests/push-api/subscribe-deny-permissions-on-prompt-expected.txt:
* LayoutTests/http/tests/push-api/subscribe-deny-permissions-on-prompt.html:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/permissions/PermissionController.h:
* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::Permissions::sourceFromContext):
(WebCore::sourceFromContext): Deleted.
* Source/WebCore/Modules/permissions/Permissions.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::queryPermission):
(WebKit::WebPageProxy::clearNotificationPermissionState):
* Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp:
(WebKit::WebPermissionControllerProxy::querySync):
* Source/WebKit/UIProcess/WebPermissionControllerProxy.h:
* Source/WebKit/UIProcess/WebPermissionControllerProxy.messages.in:
* Source/WebKit/WebProcess/Notifications/NotificationPermissionRequestManager.cpp:
(WebKit::NotificationPermissionRequestManager::startRequest):
(WebKit::NotificationPermissionRequestManager::permissionLevel): Deleted.
* Source/WebKit/WebProcess/Notifications/NotificationPermissionRequestManager.h:
* Source/WebKit/WebProcess/Notifications/WebNotificationManager.cpp:
(WebKit::WebNotificationManager::initialize):
(WebKit::WebNotificationManager::didUpdateNotificationDecision):
(WebKit::WebNotificationManager::didRemoveNotificationDecisions):
(WebKit::WebNotificationManager::removeAllPermissionsForTesting):
(WebKit::WebNotificationManager::policyForOrigin const): Deleted.
* Source/WebKit/WebProcess/Notifications/WebNotificationManager.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
(WebKit::WebNotificationClient::requestPermission):
(WebKit::WebNotificationClient::checkPermission):
(WebKit::WebNotificationClient::clearNotificationPermissionState): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
(WebKit::WebPermissionController::query):
(WebKit::WebPermissionController::querySync):
(WebKit::WebPermissionController::webPageProxyIdentifier):
(WebKit::WebPermissionController::removeEntryFromCache):
(WebKit::WebPermissionController::clearCache):
(WebKit::WebPermissionController::tryProcessingRequests):
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:
(isType):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::clearNotificationPermissionState): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::removeAllWebNotificationPermissions):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::removeAllWebNotificationPermissions):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveSynchronousMessageFromInjectedBundle):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/92a17b98d823519296e40f2440baa5af4e049c92

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87267 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31354 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18085 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96377 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149842 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91242 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29719 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/25935 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79377 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91270 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92883 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24011 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74065 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23869 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/78985 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79218 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66884 "Found 2 new API test failures: /WebKit2Gtk/TestWebKitWebView:/webkit/WebKitWebView/notification, /WebKit2Gtk/TestWebKitWebView:/webkit/WebKitWebView/notification-initial-permission-allowed") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27437 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/13054 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27388 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14069 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29073 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/36930 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29007 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33346 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/permissions/PermissionController.h:
```diff
@@ -31,14 +31,19 @@
 
 namespace WebCore {
 
+enum class QueryType : uint8_t {
```
enum class QueryType : bool {

cdumez replied:
QueryType is very vague, I suggest coming up with a better name. Maybe:
enum class Asynchronous : bool { No, Yes };

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/PermissionController.h:
```diff
@@ -31,14 +31,19 @@
 namespace WebCore {
 
+enum class QueryType : uint8_t {
+    Sync,
```
Could all be on one line.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.messages.in:
```diff
@@ -622,6 +622,8 @@ messages -> WebPageProxy {
 
     QueryPermission(struct WebCore::ClientOrigin origin, struct WebCore::PermissionDescriptor descriptor) -> (std::optional<WebCore::PermissionState> state, bool shouldCache)
 
```
Maybe no blank line here since the 2 are so related.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
```diff
@@ -119,9 +125,24 @@ NotificationClient::Permission WebNotificationClient::checkPermission(ScriptExec
-    NotificationClient::Permission resultPermission;
-    callOnMainRunLoopAndWait([&resultPermission, origin = origin->data().toString().isolatedCopy()] {
-        resultPermission = WebProcess::singleton().supplement<WebNotificationManager>()->policyForOrigin(origin);
+    NotificationClient::Permission resultPermission = NotificationClient::Permission::Default;
```
auto

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
```diff
@@ -119,9 +125,24 @@ NotificationClient::Permission WebNotificationClient::checkPermission(ScriptExec
-        resultPermission = WebProcess::singleton().supplement<WebNotificationManager>()->policyForOrigin(origin);
+    NotificationClient::Permission resultPermission = NotificationClient::Permission::Default;
+    callOnMainRunLoopAndWait([this, &resultPermission, context, origin] {
+        auto page = this->m_page->corePage();
```
No need for `this->`.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
```diff
@@ -119,9 +125,24 @@ NotificationClient::Permission WebNotificationClient::checkPermission(ScriptExec
-    callOnMainRunLoopAndWait([&resultPermission, origin = origin->data().toString().isolatedCopy()] {
-        resultPermission = WebProcess::singleton().supplement<WebNotificationManager>()->policyForOrigin(origin);
+    NotificationClient::Permission resultPermission = NotificationClient::Permission::Default;
+    callOnMainRunLoopAndWait([this, &resultPermission, context, origin] {
```
You can only use `context` on the calling thread so I don't think it is OK to capture `context` here to use it on the main thread. Get the information you need from the `context` on the calling thread, then capture only that information in the lambda (calling `crossThreadCopy()` on it since you're passing those to another thread).

Same comment for origin, SecurityOrigin is not thread to use from different thread, you should capture:
`originData = crossThreadCopy(origin->data())`

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
```diff
@@ -119,9 +125,24 @@ NotificationClient::Permission WebNotificationClient::checkPermission(ScriptExec
+        controller.query(ClientOrigin { context->topOrigin().data(), originData }, PermissionDescriptor { WebCore::PermissionName::Notifications }, WebCore::QueryType::Sync, [&resultPermission] (auto permissionState) mutable {
+
+            if (permissionState) {
+                auto permission = permissionState.value();
```
No need for this local

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
```diff
@@ -119,9 +125,24 @@ NotificationClient::Permission WebNotificationClient::checkPermission(ScriptExec
+
+            if (permissionState) {
+                auto permission = permissionState.value();
+                if (permission == WebCore::PermissionState::Granted)
```
Maybe we can `switch(*permissionState)` here.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:
```diff
@@ -39,11 +39,12 @@ class WebPermissionController final : public CanMakeWeakPtr<WebPermissionControl
 private:
     explicit WebPermissionController(WebPage&);
 
-    // WebCore::PermissionController
```
Please keep this comment here, I believe it still applies to the addObserver() / removeObserver() functions.

RupinMittal replied:
Fixed!

cdumez commented on Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
```diff
@@ -888,6 +888,7 @@ void TestRunner::denyWebNotificationPermissionOnPrompt(JSStringRef origin)
 void TestRunner::removeAllWebNotificationPermissions()
 {
     WKBundleRemoveAllWebNotificationPermissions(InjectedBundle::singleton().bundle(), page());
+    postSynchronousPageMessage("ResetNotificationPermissions");
```
Can we use `RemoveAllWebNotificationPermissions` to match the name of the function? I believe this is the pattern in our existing code and it makes it clearer what caused this IPC.

RupinMittal replied:
Fixed!

cdumez commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -794,6 +794,11 @@ bool TestController::denyNotificationPermissionOnPrompt(WKStringRef originString
     return true;
 }
 
+void TestController::resetNotificationPermissions()
```
Should probably be named removeAllWebNotificationPermissions.

RupinMittal replied:
Fixed!

szewai commented on Source/WebCore/Modules/permissions/PermissionController.h:
```diff
@@ -31,14 +31,16 @@
 public:
     virtual ~PermissionController() = default;
-    virtual void query(WebCore::ClientOrigin&&, PermissionDescriptor&&, CompletionHandler<void(std::optional<PermissionState>)>&&) = 0;
+    virtual void query(WebCore::ClientOrigin&&, PermissionDescriptor&&, Asynchronous&&, CompletionHandler<void(std::optional<PermissionState>)>&&) = 0;
```
No need to use && with Asynchronous

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
```diff
@@ -119,9 +125,27 @@ NotificationClient::Permission WebNotificationClient::checkPermission(ScriptExec
+    auto resultPermission = NotificationClient::Permission::Default;
+    auto topOrigin = context->topOrigin().data();
+    callOnMainRunLoopAndWait([this, &resultPermission, topOrigin = crossThreadCopy(context->topOrigin().data()), securityOrigin = crossThreadCopy(origin->data())] {
+        auto page = m_page->corePage();
```
Can't m_page be null? It seems this change would break our support for Notification.permission in ServiceWorker (@cdumez and @youennf can confirm), because ServiceWorkerGlobalScope does not have PermissionController, and it has NotificationClient.

cdumez replied:
Yes, I believe theoretically that `m_page` can be null.

It is a good point that we should be careful about not breaking Notification.permission support in service workers. I would have expected us to have tests for this though.

cdumez replied:
Those tests should be under http/tests/workers/service. Looks like we do have some coverage.

cdumez replied:
Actually, none of the http/tests/workers/service tests seems to call `Notification.permission` *from* the service worker, only from the main page. I suspect we don't have test coverage for `Notification.permission` in service workers and like @szewai I suspect it is broken by your patch since service workers don't have a WebPage.

Sadly, I don't think it will be easy to fix until we have support for the generic Permissions.query() in (service) workers :( Permissions.query() currently requires a WebPage to function and that will probably have to change to support service workers.

RupinMittal replied:
Two things:
1. Do we think the implementation before this patch would work for serviceWorkers? It yes, then maybe we can use that implementation if `m_page` is `null` and if not, then we use the code path in this patch. I could add an extra check in the code for these 2 cases.
2. The only reason `m_page` is used is to access the `WebPermissionController` object. Is there a way to access the controller without getting it from the page? If yes, then we can remove the use of `m_page`, get the controller a different way and then this implementation could work for service workers as well correct?

cdumez replied:
Currently the service workers don't have a WebPermissionController since it is constructed by the WebPage and service workers don't have an associated WebPage (only a dummy WebCore::Page).

Yes, I think the implementation before this patch should work for service workers, though I wish we had layout tests coverage to confirm. Using the old code path when m_page is null wouldn't really be useful, would it? We would still need to maintain the old code path, just for service workers (with very little or no test coverage of that code path). The whole point of this patch was to have a single code path and get rid of the WebProcess cache I think.

RupinMittal replied:
We put this patch on hold since there was no support for service workers in Permissions::query. Since there is support now (after https://github.com/WebKit/WebKit/pull/3570), we can do the refactoring now.

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -52,8 +53,18 @@ void WebPermissionController::query(WebCore::ClientOrigin&& origin, WebCore::Per
+        m_page->sendSync(Messages::WebPageProxy::QueryPermissionSync(origin, descriptor), Messages::WebPageProxy::QueryPermissionSync::Reply(state, shouldCache));
+        
+        completionHandler(state);
+        if (shouldCache && state)
```
IIUC, the result is never cached and we will send a sync message to UI process per request. I think this is not ideal and we can have better implementation after we support permission change from UI process, i.e. caching the result and let UI process push updates; we may even eliminate the need for sync message and just return default when cache is empty (e.g. if page has not called query/requestPermission). 
So it's probably a better idea to implement change event first.

cdumez replied:
I think the sync IPC is acceptable for now, in order to have a single code path for both the sync and async APIs. Otherwise, some of Rupin's work ends up getting duplicated.

szewai replied:
I would expect the duplicate work to be little after what Rupin has already done (like the permission requester cache), at least in Worker support and change event support, and we would have more flexibility in implementation (e.g. using async message instead of sync message) with these support, so I think it's better getting those two items done first.

cdumez replied:
Ok, I think Rupin wanted to do this patch because he wanted to move his requester cache from the WebPageProxy to the WebsiteDataStore. Right now the cache used by the old code path in the WebProcess would cause inconsistency between the sync and async APIs.

Anyway, I guess we can keep the requester cache on the WebPageProxy for now to match the WebContent process side requester cache that is used by the legacy sync API.

szewai reviewed (changes requested):
I think this needs to be be done after adding Worker support for Permissions API, and is better be done after change event is supported (or making web process able to cache the state to avoid sync messages).

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -57,8 +56,18 @@ void WebPermissionController::query(WebCore::ClientOrigin&& origin, WebCore::Per
+        WebProcess::singleton().sendSync(Messages::WebPermissionControllerProxy::QuerySync(origin, descriptor, proxyIdentifier, source), Messages::WebPermissionControllerProxy::QuerySync::Reply(state, shouldCache));
+        completionHandler(state);
+        if (shouldCache && state)
+            updateCache(origin, descriptor, *state);
```
I think updateCache should be put before completionHandler, because technically completionHandler should be called after query operation, and client can change the state of WebPermissionController in completionHandler.

To avoid such issue, I would suggest adding a querySync that returns a permissionState, instead of changing the existing query(). E.g.
std::optional<WebCore::PermissionState> querySync(ClientOrigin&&, PermissionDescriptor&&, Page*, PermissionQuerySource)

RupinMittal replied:
I agree, updateCache cache should come before the completion handler as well. I'll change this in `WebPermissionController::tryProcessingRequests` as well. But I feel that we can do this and avoid the issue without adding a new function since that would duplicate code right?

szewai replied:
The querySync is to make it clear for the caller that the function will return sync-ly; we usually don't let caller pass CompletionHandler for pure sync operation (the WebCore::Synchronous mode).
For the duplicate code, we can add helper functions to share code between two functions, but I don't think the duplicate part is that much (query cache and get identifier). Given that the cache is not actually in use and may get removed, it may be even less.

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
```diff
@@ -102,41 +111,46 @@ void WebNotificationClient::requestPermission(ScriptExecutionContext& context, P
-    callOnMainRunLoopAndWait([&resultPermission, origin = origin->data().toString().isolatedCopy()] {
-        resultPermission = WebProcess::singleton().supplement<WebNotificationManager>()->policyForOrigin(origin);
+    Page* page = nullptr;
+    if (source == WebCore::PermissionQuerySource::Window || source == WebCore::PermissionQuerySource::DedicatedWorker)
```
m_page is null for ServiceWorker, so this can be:
`Page* page = m_page ? m_page->corePage() : nullptr;`

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
```diff
@@ -102,41 +111,46 @@ void WebNotificationClient::requestPermission(ScriptExecutionContext& context, P
+
+    auto resultPermission = NotificationClient::Permission::Default;
+    callOnMainRunLoopAndWait([&resultPermission, topOrigin = crossThreadCopy(context->topOrigin().data()), securityOrigin = crossThreadCopy(context->securityOrigin()->data()), page, source = *source] {
+        PermissionController::shared().query(ClientOrigin { topOrigin, securityOrigin }, PermissionDescriptor { WebCore::PermissionName::Notifications }, page, source, WebCore::Asynchronous::No, [&resultPermission] (auto permissionState) mutable {
```
Add a querySync to PermissionController and change this to:
`auto permissionState = PermissionController::shared().querySync(...);`

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
```diff
@@ -102,41 +111,46 @@ void WebNotificationClient::requestPermission(ScriptExecutionContext& context, P
+    auto resultPermission = NotificationClient::Permission::Default;
+    callOnMainRunLoopAndWait([&resultPermission, topOrigin = crossThreadCopy(context->topOrigin().data()), securityOrigin = crossThreadCopy(context->securityOrigin()->data()), page, source = *source] {
+        PermissionController::shared().query(ClientOrigin { topOrigin, securityOrigin }, PermissionDescriptor { WebCore::PermissionName::Notifications }, page, source, WebCore::Asynchronous::No, [&resultPermission] (auto permissionState) mutable {
+            if (permissionState) {
```
```
if (!permissionState)
    return;
```

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
```diff
@@ -102,41 +111,46 @@ void WebNotificationClient::requestPermission(ScriptExecutionContext& context, P
+                case WebCore::PermissionState::Denied:
+                    resultPermission = NotificationClient::Permission::Denied;
+                    break;
+                default: 
```
The only default case is PermissionState::Prompt so you may just use PermissionState::Prompt for clarity and set resultPermission accordingly.

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
```diff
@@ -102,41 +111,46 @@ void WebNotificationClient::requestPermission(ScriptExecutionContext& context, P
+                    resultPermission = NotificationClient::Permission::Denied;
+                    break;
+                default: 
+                    break;
```
Not need to break for the last label

szewai reviewed (changes requested):
Test failure looks relevant, please fix them

RupinMittal committed:
Refactor Notification::permission to follow the same code path as Permissions::query
https://bugs.webkit.org/show_bug.cgi?id=243502

Reviewed by NOBODY (OOPS!).

There are two functions for obtaining the permission state of the Notifications API.
We need to provide support for both to ensure compatibility with web standards. This
patch ensures that they follow the same code path since having two different paths is
unnecessary and hurts maintainability. This code path goes to the UIProcess and then
to the client to get the permission state.

This patch also removes the now uneeded cache in WebNotificationManager that
Notification::permission was relying on and enables the WebProcess cache in
WebPermissionController to be used.

* LayoutTests/http/tests/notifications/service-worker-notification-permission-expected.txt: Added.
* LayoutTests/http/tests/notifications/service-worker-notification-permission.html: Added.
* LayoutTests/http/tests/notifications/service-worker-notification-permission.js: Added.
(self.onmessage):
* LayoutTests/http/tests/push-api/subscribe-deny-permissions-on-prompt-expected.txt:
* LayoutTests/http/tests/push-api/subscribe-deny-permissions-on-prompt.html:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/permissions/PermissionController.h:
* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::Permissions::sourceFromContext):
(WebCore::sourceFromContext): Deleted.
* Source/WebCore/Modules/permissions/Permissions.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::queryPermission):
(WebKit::WebPageProxy::clearNotificationPermissionState):
* Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp:
(WebKit::WebPermissionControllerProxy::querySync):
* Source/WebKit/UIProcess/WebPermissionControllerProxy.h:
* Source/WebKit/UIProcess/WebPermissionControllerProxy.messages.in:
* Source/WebKit/WebProcess/Notifications/NotificationPermissionRequestManager.cpp:
(WebKit::NotificationPermissionRequestManager::startRequest):
(WebKit::NotificationPermissionRequestManager::permissionLevel): Deleted.
* Source/WebKit/WebProcess/Notifications/NotificationPermissionRequestManager.h:
* Source/WebKit/WebProcess/Notifications/WebNotificationManager.cpp:
(WebKit::WebNotificationManager::initialize):
(WebKit::WebNotificationManager::didUpdateNotificationDecision):
(WebKit::WebNotificationManager::didRemoveNotificationDecisions):
(WebKit::WebNotificationManager::removeAllPermissionsForTesting):
(WebKit::WebNotificationManager::policyForOrigin const): Deleted.
* Source/WebKit/WebProcess/Notifications/WebNotificationManager.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
(WebKit::WebNotificationClient::requestPermission):
(WebKit::WebNotificationClient::checkPermission):
(WebKit::WebNotificationClient::clearNotificationPermissionState): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
(WebKit::WebPermissionController::query):
(WebKit::WebPermissionController::querySync):
(WebKit::WebPermissionController::webPageProxyIdentifier):
(WebKit::WebPermissionController::removeEntryFromCache):
(WebKit::WebPermissionController::clearCache):
(WebKit::WebPermissionController::tryProcessingRequests):
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:
(isType):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::clearNotificationPermissionState): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::removeAllWebNotificationPermissions):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::removeAllWebNotificationPermissions):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveSynchronousMessageFromInjectedBundle):

szewai commented on LayoutTests/http/tests/push-api/subscribe-deny-permissions-on-prompt.html:
```diff
@@ -16,35 +16,30 @@
-                reject('Test requires internals.');
-                return;
-            }
+        if (!window.internals)
```
Can you explain in commit message that why this change is needed? It's not clear to me why pushManager would return different results after this patch

RupinMittal replied:
`WebNotificationClient::checkPermission` (and thus, `Notification::permission`) used to get the permission state from a cache in `WebNotificationManager`. This cache did not get updated by calling `TestController::denyNotificationPermissionOnPrompt` and then `WebNotificationClient::requestPermission` (which is what happens in this test). So when request2 occurred, `WebNotificationClient::checkPermission` would return default, then fail the transient activation check and reject the promise with a "user gesture" error. 

This patch removes this cache and refactors `WebNotificationClient::checkPermission` (and thus, `Notification::permission`) to get the permission state from the client (which is `TestController` in the case of this test). The cache in `TestController` that stores the permissions state does get updated by calling `TestController::denyNotificationPermissionOnPrompt` and then `WebNotificationClient::requestPermission`. So now, when request2 occurs, `WebNotificationClient::checkPermission` returns denied and rejects the promise with a "user denied push permission" error before reaching the transient activation check.

I spoke with @bnham about this and learned that the point of this layout test is to check that given a user gesture, the first call to `PushManager::subscribe` would consume the transient activation and that the second call would not have a transient activation. Due to the change described above, this check on the second call was being circumvented. We can ensure that this check occurs by altering the test the way I did in this patch.

You're right, the reason for change is not clear just by looking at the test, so I'll update the commit message with the above explanation once the bots finish and I make any other suggested changes.

szewai replied:
From what's described above, It seems the right thing is to update WebKitTestRunner, making it cache the request result like a real client would do, instead of changing the test, since you don't intend to change the tested behavior (that first subscribe request fails due to user activation and second  fails due to permission request).

Our current design for permissions is that each query and request goes to client, which means client is responsible to cache the result if needed. 
Specifically, you might call `setPermission` on `WebNotificationProvider` in `TestController::decidePolicyForNotificationPermissionRequest.` (to simulate user denies the prompt and browser remembers the choice).

RupinMittal replied:
I believe WebKitTestRunner already caches the result. `TestController::decidePolicyForNotificationPermissionRequest.` already calls `setPermission` on `WebNotificationProvider`. denyOnPrompt means that if you request/prompt for permission, the state will be set to denied. The test was expecting that after denyOnPrompt is called, then first request does a prompt, the second request will get that the permission state is default? But that doesn't seem to make sense. The permission state should be denied, no? The reason that the test was passing before this patch was because it was relying on the fact that WebKitTestRunner wasn't updating the cache in WebNotificationManager. Now that this cache is gone, and checkPermission actually goes to the client, and testRunner returns the result based on its cache in WebNotificationProvider, the second request correctly gets back that the permission state is denied.

szewai commented on Source/WebKit/WebProcess/Notifications/WebNotificationManager.cpp:
```diff
@@ -73,7 +80,12 @@ WebNotificationManager::~WebNotificationManager()
 {
 #if ENABLE(NOTIFICATIONS)
-    m_permissionsMap = parameters.notificationPermissions;
+    auto& webPermissionController = downcast<WebPermissionController>(PermissionController::shared());
+    for (auto const& [originString, allowed] : parameters.notificationPermissions) {
+        auto topOrigin = SecurityOrigin::createFromString(originString)->data();
+        auto permissionState = allowed ? PermissionState::Granted : PermissionState::Denied;
+        webPermissionController.updateCache(ClientOrigin { topOrigin, SecurityOriginData { } }, PermissionDescriptor { PermissionName::Notifications }, permissionState);
+    }
```
Why this change? I thought we only want one implementation for both APIs, i.e. we will only use queryPermission to pull notification's permission.
This makes the cache dependent on WKNotificationManager: e.g. if client implements queryPermission but not WKNotificationManager, the result will be wrong (because cache will not be updated).

RupinMittal replied:
This patch contains other code changes that ensure that calling `queryPermission` will also update the cache (if the permission state being queried is the state of the Notifications API.

The reason for this change was that in a previous version of this patch, the EWS bots detected a API test failure on GTK. The failure was that `Notification::permission` was incorrectly returning "default" instead of "granted". It turns out that the test client did not implement `queryPermission` and the test had been succeeding because of the cache in `WebNotificationManager`. Since this patch removes that cache and refactors `Notification::permission` to use `queryPermission`, the result was coming back as default. There were two options to fix this:

1. Implement `queryPermission` in the GTK test client
2. Make the not-working WebProcess cache in `WebPermissionController` work for notifications. 

The second option seemed easier and more beneficial since we'll have to do this for the on change event later on anyways. I wasn't sure how to test if this will fix the API test since it was a GTK failure, so I guess we'll have to wait for the bots to check. :(

szewai replied:
> This patch contains other code changes that ensure that calling queryPermission will also update the cache (if the permission state being queried is the state of the Notifications API.

Let's say client implements queryPermission but not WKNotificationManager, and user sets allowed notification (e.g. in per-site settings in Safari) for a site.
For the first permission::query, client returns allowed, and WebPermissionController caches the result. Then user decides to change the setting to denied (no longer wants to receive notification from the site), and tells WebKit client. For the second permission query, what would be returned with your patch?

IIUC, before this patch, notification.permission relies on WKNotificationManager and permission.query relies on queryPermission. After this patch, both rely on either WKNotificationManager or queryPermission + WKNotificationManager, which does not seem like a progress to me.

> The second option seemed easier and more beneficial since we'll have to do this for the on change event later on anyways.

The reason we implement this before change event is because we think this change would work on its own. If this change relies on cache to work, then I think we should implement change event first,  since the cache might not be in this form (an origin-to-permission map). You can imagine having WebPermissionController track PermissionStatus (or PermissionObserver) objects which hold the state, then you will need to revamp the cache.

cdumez commented on Source/WebKit/WebProcess/Notifications/NotificationPermissionRequestManager.cpp:
```diff
@@ -85,7 +91,9 @@ void NotificationPermissionRequestManager::startRequest(const SecurityOriginData
-            WebProcess::singleton().supplement<WebNotificationManager>()->didUpdateNotificationDecision(securityOrigin.toString(), allowed);
+            auto permissionState = allowed ? PermissionState::Granted : PermissionState::Denied;
+            auto& webPermissionController = downcast<WebPermissionController>(PermissionController::shared());
+            webPermissionController.updateCache(ClientOrigin { securityOrigin, SecurityOriginData { } }, PermissionDescriptor { PermissionName::Notifications }, permissionState);
```
This seems wrong, how it is correct to have a ClientOrigin with a null SecurityOriginData() inside?

cdumez commented on Source/WebKit/WebProcess/Notifications/WebNotificationManager.cpp:
```diff
@@ -73,7 +80,12 @@ WebNotificationManager::~WebNotificationManager()
+    for (auto const& [originString, allowed] : parameters.notificationPermissions) {
+        auto topOrigin = SecurityOrigin::createFromString(originString)->data();
+        auto permissionState = allowed ? PermissionState::Granted : PermissionState::Denied;
+        webPermissionController.updateCache(ClientOrigin { topOrigin, SecurityOriginData { } }, PermissionDescriptor { PermissionName::Notifications }, permissionState);
```
Same comment, your ClientOrigin seems wrong.

cdumez commented on Source/WebKit/WebProcess/Notifications/WebNotificationManager.cpp:
```diff
@@ -82,7 +94,10 @@ void WebNotificationManager::initialize(const WebProcessCreationParameters& para
+    auto& webPermissionController = downcast<WebPermissionController>(PermissionController::shared());
+    auto topOrigin = SecurityOrigin::createFromString(originString)->data();
+    auto permissionState = allowed ? PermissionState::Granted : PermissionState::Denied;
+    webPermissionController.updateCache(ClientOrigin { topOrigin, SecurityOriginData { } }, PermissionDescriptor { PermissionName::Notifications }, permissionState);
```
Again, bad ClientOrigin.

cdumez commented on Source/WebKit/WebProcess/Notifications/WebNotificationManager.cpp:
```diff
@@ -92,33 +107,19 @@ void WebNotificationManager::didUpdateNotificationDecision(const String& originS
+    auto& webPermissionController = downcast<WebPermissionController>(PermissionController::shared());
     for (auto& originString : originStrings)
-        m_permissionsMap.remove(originString);
+        webPermissionController.removeEntryFromCache(ClientOrigin { SecurityOrigin::createFromString(originString)->data(), SecurityOriginData { } });
```
Also bad.

cdumez commented on Source/WebKit/WebProcess/Notifications/WebNotificationManager.cpp:
```diff
@@ -92,33 +107,19 @@ void WebNotificationManager::didUpdateNotificationDecision(const String& originS
 #if ENABLE(NOTIFICATIONS)
-    m_permissionsMap.clear();
+    auto& webPermissionController = downcast<WebPermissionController>(PermissionController::shared());
+    webPermissionController.clearCache();
```
Well, this seems wrong isn't it? We used to clear only notification permissions, now we clear ALL permissions?

RupinMittal replied:
Yeah, this is a mistake. I'll fix this in the next upload of the patch.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
```diff
@@ -102,41 +111,43 @@ void WebNotificationClient::requestPermission(ScriptExecutionContext& context, P
-    NotificationClient::Permission resultPermission;
-    callOnMainRunLoopAndWait([&resultPermission, origin = origin->data().toString().isolatedCopy()] {
-        resultPermission = WebProcess::singleton().supplement<WebNotificationManager>()->policyForOrigin(origin);
+    auto* page = m_page ? m_page->corePage() : nullptr;
```
m_page is a main thread object and is thus not safe to access off the main thread like you're doing here.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
```diff
@@ -102,41 +111,43 @@ void WebNotificationClient::requestPermission(ScriptExecutionContext& context, P
+        if (!permissionState)
+            return;
+
+        switch (*permissionState) {
```
Seems like we should have a helper function to convert one enum type to the other.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:
```diff
@@ -74,3 +79,7 @@ class WebPermissionController final : public CanMakeWeakPtr<WebPermissionControl
 } // namespace WebCore
+
+SPECIALIZE_TYPE_TRAITS_BEGIN(WebKit::WebPermissionController)
+    static bool isType(const WebCore::PermissionController&) { return true; }
```
How could this possibly be correct? You're doing zero type validation?

cdumez commented on LayoutTests/http/tests/push-api/subscribe-deny-permissions-on-prompt-expected.txt:
```diff
@@ -1,9 +1,8 @@
 PASS: document subscribe without user gesture was error: NotAllowedError
-PASS: document subscribe with user gesture was error: NotAllowedError
-PASS: document subscribe with consumed user gesture failed with user gesture error
+PASS: document subscribe with user gesture consumed the user gesture
```
The difference in behavior here is worrisome. This is meant to be a refactoring that maintain pre-existing behavior. Mixing behavior changes and refactoring in the same patch is usually a recipe for having a bad time later.
